db/crud.go

```diff
@@ -814,6 +814,235 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
+				base.InfofCtx(db.Ctx, base.KeyCache, "updateDoc %q: Unused sequence #%d", base.UD(docid), docSequence)
+				unusedSequences = append(unusedSequences, docSequence)
+			}
+
+			for {
+				if docSequence, err = db.sequences.nextSequence(); err != nil {
+					return unusedSequences, err, docSequence
+				}
+
+				if docSequence > doc.Sequence {
+					break
+				} else {
+					releaseErr := db.sequences.releaseSequence(docSequence)
+					if releaseErr != nil {
+						base.Warnf(base.KeyCRUD, "Error returned when releasing sequence %d. Falling back to skipped sequence handling.  Error:%v", docSequence, err)
+					}
+				}
+			}
+			// Could add a db.Sequences.nextSequenceGreaterThan(doc.Sequence) to push the work down into the sequence allocator
+			//  - sequence allocator is responsible for releasing unused sequences, could optimize to do that in bulk if needed
+		}
+	}
+	doc.Sequence = docSequence
+	doc.UnusedSequences = unusedSequences
+
+	// The server TAP/DCP feed will deduplicate multiple revisions for the same doc if they occur in
+	// the same mutation queue processing window. This results in missing sequences on the change listener.
+	// To account for this, we track the recent sequence numbers for the document.
+	if doc.RecentSequences == nil {
+		doc.RecentSequences = make([]uint64, 0, 1+len(unusedSequences))
+	}
+
+	if len(doc.RecentSequences) >= kMaxRecentSequences {
+		// Prune recent sequences that are earlier than the nextSequence.  The dedup window
+		// on the feed is small - sub-second, so we usually shouldn't care about more than
+		// a few recent sequences.  However, the pruning has some overhead (read lock on nextSequence),
+		// so we're allowing more 'recent sequences' on the doc (20) before attempting pruning
+		stableSequence := db.changeCache.GetStableSequence(docid).Seq
+		count := 0
+		for _, seq := range doc.RecentSequences {
+			// Only remove sequences if they are higher than a sequence that's been seen on the
+			// feed. This is valid across SG nodes (which could each have a different nextSequence),
+			// as the mutations that this node used to rev nextSequence will at some point be delivered
+			// to each node.
+			if seq < stableSequence {
+				count++
+			} else {
+				break
+			}
+		}
+		if count > 0 {
+			doc.RecentSequences = doc.RecentSequences[count:]
+		}
+	}
+
+	// Append current sequence and unused sequences to recent sequence history
+	doc.RecentSequences = append(doc.RecentSequences, unusedSequences...)
+	doc.RecentSequences = append(doc.RecentSequences, docSequence)
+
+	return unusedSequences, nil, docSequence
+}
+
+// Bool is for bail out
+func (db *Database) updateRevTree(doc *Document, prevCurrentRev, newRevID, docid string, changedPrincipals []string, changedUsers []string, access channels.AccessMap, roles channels.AccessMap, channelSet base.Set, syncExpiry *uint32, oldBody string) (base.Set, channels.AccessMap, channels.AccessMap, *uint32, string, bool, []string, []string, error) {
```

Comment:
Can we name the return values. It's impossible to tell what this function is returning from just reading the function signature.